### PR TITLE
Disallow the use of value source providers at configuration time unless explicitly opted-in

### PR DESCRIPTION
**Pull request by bamboo**

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/api/provider/Provider.java [hunk 1]
```diff
@@ -144,4 +144,12 @@
...
     @Incubating
     Provider<T> orElse(Provider<? extends T> provider);
+
+    /**
+     * Returns a view of this {@link Provider} which can be safely read at configuration time.
+     *
+     * @since 6.5
+     */
+    @Incubating
+    Provider<T> forUseAtConfigurationTime();
```
Comment (adammurdoch) [inline subprojects/core-api/src/main/java/org/gradle/api/provider/Provider.java:154]: should we define what happens if this is not possible? for example , when the provider value is derived from a task output and is not available until the task has completed?

**Comment by bamboo** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/api/provider/Provider.java (hunk 1 above)
Comment (bamboo) [reply subprojects/core-api/src/main/java/org/gradle/api/provider/Provider.java]: added follow-up issue: https://github.com/gradle/instant-execution/issues/329

**Comment by bamboo** [comment]
@bot-gradle test this

**Comment by bamboo** [comment]
@bot-gradle test this

**Comment by bamboo** [comment]
@bot-gradle test this

**Comment by bamboo** [comment]
@bot-gradle test this

**Comment by bamboo** [comment]
@bot-gradle test RFN

**Pull request merged**